package/setup.py

```diff
@@ -130,33 +129,93 @@ def get(self, option_name, default=None):
         except configparser.NoOptionError:
             return default
 
-class MDAExtension(Extension, object):
-    """Derived class to cleanly handle setup-time (numpy) dependencies.
+class MDADistribution(Distribution, object):
+    """Derived class to cleanly handle setup-time (cython/numpy) dependencies.
     """
-    # The only setup-time numpy dependency comes when setting up its
-    #  include dir.
-    # The actual numpy import and call can be delayed until after pip
-    #  has figured it must install numpy.
-    # This is accomplished by passing the get_numpy_include function
-    #  as one of the include_dirs. This derived Extension class takes
-    #  care of calling it when needed.
-    def __init__(self, *args, **kwargs):
-        self._mda_include_dirs = []
-        super(MDAExtension, self).__init__(*args, **kwargs)
-
-    @property
-    def include_dirs(self):
-        if not self._mda_include_dirs:
-            for item in self._mda_include_dir_args:
-                try:
-                    self._mda_include_dirs.append(item()) #The numpy callable
-                except TypeError:
-                    self._mda_include_dirs.append(item)
-        return self._mda_include_dirs
-
-    @include_dirs.setter
-    def include_dirs(self, val):
-        self._mda_include_dir_args = val
+    # Setup-time cython dependency comes when cythonizing sources (cython might
+    #  not be available) or configuring numpy's include dir (numpy might
+    #  not be available either).
+    # The actual cython/numpy imports and calls can be delayed until after
+    #  pip/setuptools have figured they must install them.
+    # This is accomplished by using a distutils.core.Distribution subclass
+    #  that lazily fills-in cython and numpy information by hooking into
+    #  the run_commands method.
+    def __init__(self, attrs=None):
+        # Attributes must exist if they're to be imported from attrs.
+        self.mda_use_cython = is_dev
+        self.mda_build_requires = None
+        super(MDADistribution, self).__init__(attrs)
+
+    def run_commands(self):
+        # The following seems fragile...
```

Comment:
Yeah I think Max is right here from what I can tell. Can we not fix this
upstream?

On Fri, 25 Mar 2016 11:04 kain88-de, <notifications@github.com> wrote:

> In package/setup.py
> <https://github.com/MDAnalysis/mdanalysis/pull/799#discussion_r57436127>:
>
> > +    # Setup-time cython dependency comes when cythonizing sources (cython might
> > +    #  not be available) or configuring numpy's include dir (numpy might
> > +    #  not be available either).
> > +    # The actual cython/numpy imports and calls can be delayed until after
> > +    #  pip/setuptools have figured they must install them.
> > +    # This is accomplished by using a distutils.core.Distribution subclass
> > +    #  that lazily fills-in cython and numpy information by hooking into
> > +    #  the run_commands method.
> > +    def __init__(self, attrs=None):
> > +        # Attributes must exist if they're to be imported from attrs.
> > +        self.mda_use_cython = is_dev
> > +        self.mda_build_requires = None
> > +        super(MDADistribution, self).__init__(attrs)
> > +
> > +    def run_commands(self):
> > +        # The following seems fragile...
>
> I don't like this because we are reinventing the wheel and circumventing
> existing solutions! I don't really like to add such a hack that is easy to
> break and not easy to understand for other people.
>
> What is done first install our install_requires or setup_requires? Also
> is MDAnalysis first build and then everything is installed?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/799/files/f3f5a69644ab0ef51ddda8f0ba74d10505013199#r57436127>
>